### ISSUE TITLE
Reflector (WebGLRenderTarget): Fog Color problem in sRGB outputEncoding

### ISSUE THREAD

**Issue by takahirox** [mrdoob/three.js#24362]
From https://github.com/mozilla/hubs/issues/5572

**Describe the bug**

Fog color in `Reflector` (`WebGLRenderTarget`) can mismatch the one in the screen.

I made an online example to show the problem.

https://raw.githack.com/takahirox/three.js/MirrorFogExample/examples/webgl_mirror_fog.html

**Screenshots**

sRGB output encoding + fog. Fog color in the reflector looks brighter than the scene. (See the green wall.)

![image](https://user-images.githubusercontent.com/7637832/179628595-d1d35be0-892e-4a6c-8bc4-87b8f9f00411.png)

sRGB output encoding + no fog. Color in the reflector looks the same as the scene's.

![image](https://user-images.githubusercontent.com/7637832/179628919-9d9ba6fb-f1e9-4ab7-918b-ff678b49d146.png)

Linear output encoding + fog. Fog color in the reflector looks the same as the scene's.

![image](https://user-images.githubusercontent.com/7637832/179629040-070f2049-842d-4154-92ea-7a3e906c711c.png)

***Live example***

https://raw.githack.com/takahirox/three.js/MirrorFogExample/examples/webgl_mirror_fog.html

**Expected behavior**

Fog color in the reflector looks same as the scene's even in sRGB outputEncoding.

**Screenshots**

If applicable, add screenshots to help explain your problem (drag and drop the image).

**Platform:**

 - Device: Desktop
 - OS: Windows 10
 - Browser: Chrome
 - Three.js version: dev

**Comment by takahirox** [comment]
Probably the root issue is color space conversion.

First rendering the scene including the fog to the reflector render target. And then rendering the scene including the reflector render target texture to the screen.

Three.js does color space conversion in shader when rendering to the scene.

https://github.com/mrdoob/three.js/blob/b90f614fd59e1983d59dc24c1aa05c9c2c3886cd/src/renderers/shaders/ShaderChunk/encodings_fragment.glsl.js

The fog color in the reflector is converted in shader. But please see the shader codes in `ShaderLib` directory. It first applys `<encodings_fragment>` and then `<fog_fragment>`. So, the fog color in the main scene is not converted.

https://github.com/mrdoob/three.js/blob/12f550e0f44f68e5b6c946dd242ee08e42d209e4/src/renderers/shaders/ShaderLib/meshphysical.glsl.js#L210-L211

Therefore, the fog colors in reflector and in the main scene can mismatch.

An easy solution may be swap the order of `<encodings_fragment>` and `<fog_fragment>`. Is there any reasons why applying fog after the color space conversion to the screen?

**Comment by WestLangley** [comment]
> Is there any reasons why applying fog after the color space conversion?

I believe [this](https://github.com/mrdoob/three.js/issues/15382#issuecomment-444938863) explains the history, although at the moment, I am unable to locate the original discussion on the topic.

**Comment by LeviPesin** [comment]
Related: https://github.com/mrdoob/three.js/pull/23937

**Comment by takahirox** [comment]
Thanks @WestLangley @LeviPesin for pointing out the related threads.

**Comment by LeviPesin** [comment]
Is this still an issue after https://github.com/mrdoob/three.js/pull/23937?

**Comment by takahirox** [comment]
Unfortunately the problem doesn't seem to be resolved yet

**Comment by donmccurdy** [comment]
Note that https://github.com/mrdoob/three.js/pull/23937 will require setting `ColorManagement.legacyMode = false` to have any effect. But I think there are other things going on here, this is tricky. 😕

**Comment by takahirox** [comment]
`ColorManagement.legacyMode = false`

![image](https://user-images.githubusercontent.com/7637832/201004697-5d9e312d-806a-4363-b019-50abbfa6ad50.png)

https://raw.githack.com/takahirox/three.js/MirrorFogExample/examples/webgl_mirror_fog.html

**Comment by takahirox** [comment]
As I commented https://github.com/mrdoob/three.js/issues/24362#issuecomment-1188411765 , swapping the order of `<encodings_fragment>` and `<fog_fragment>` (and no fog color space conversion in CPU) seems to resolve the problem.

![image](https://user-images.githubusercontent.com/7637832/201416879-cd6d4192-2258-404f-9d35-e80c0c0da937.png)

https://raw.githack.com/takahirox/three.js/MirrorFogExample2/examples/webgl_mirror_fog.html

https://github.com/takahirox/three.js/commit/4788ed5e5b1dc33a4c045e1ebfc7114bd77b2933

If I understand correctly, the reason why the current order `<encodings_fragment>` and then `<fog_fragment>` was to match fog color and clear color. But with #23937 we do clear color space conversion so it may be ok to swap the order now?

**Comment by WestLangley** [comment]
Fog should be applied in scene-referred linear color space.

Since tone mapping converts from scene-referred color space to display-referred color space, the fog chunk should _precede_ tone mapping.

```glsl
#include <fog_fragment>
#include <tonemapping_fragment>
#include <encodings_fragment>
```

Ideally, we should adhere to that convention -- and make whatever changes are required to accommodate it.

/ping @donmccurdy (color management)
/ping @bhouston (legacy conventions)

**Comment by donmccurdy** [comment]
> Fog should be applied in scene-referred linear color space.
> 
> ...
> ```
> #include <fog_fragment>
> #include <tonemapping_fragment>
> #include <encodings_fragment>
> ```

Yes, agreed. This would be simpler and easier to reason about than the current ordering. For implementation purposes it would be clear improvement, and makes bugs like this easier to address.

The troublesome part is breaking convention, and explaining that to users. Consider a user writing this today:

```javascript
// (A) ColorManagement.legacyMode = true
renderer.outputEncoding = sRGBEncoding;
scene.fog.color.setRGB( 40, 40, 40 );              // sRGB → Linear-sRGB 🚨
scene.background.setRGB( 40, 40, 40 );             // sRGB
renderer.setClearColor( new Color( 40, 40, 40 ) ); // sRGB

material.color.setRGB( 40, 40, 40 );               // Linear-sRGB
light.color.setRGB( 40, 40, 40 );                  // Linear-sRGB

...

// (B) ColorManagement.legacyMode = false
renderer.outputEncoding = sRGBEncoding;
scene.fog.color.setRGB( 40, 40, 40 );              // Linear-sRGB
scene.background.setRGB( 40, 40, 40 );             // Linear-sRGB
renderer.setClearColor( new Color( 40, 40, 40 ) ); // Linear-sRGB

material.color.setRGB( 40, 40, 40 );               // Linear-sRGB
light.color.setRGB( 40, 40, 40 );                  // Linear-sRGB
```

Under `legacyMode = true` (A) it's awkward. Fog is related to background and clear color — none are affected by lighting, and you probably want them all to match — but now they're specified in different color spaces.

Under `legacyMode = false` (B) this should not be a breaking change. We still assume all RGB components are Linear-sRGB, and convert them internally as needed, when setting background or clear color. Only the internal conversion would change for fog.

***

Most users (and examples) are currently in case (A). I wish there were a gentler way to make the change under that scenario, but I'm not sure there is.

**Comment by donmccurdy** [comment]
^My concerns above are addressed in r152, as color management is now enabled by default.

Discussion continued in https://github.com/mrdoob/three.js/pull/26208.
